contracts/bonding/BondingManager.sol

```diff
@@ -306,21 +309,63 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
 
+        uint256 currentRound = roundsManager().currentRound();
+
         Transcoder storage t = transcoders[_transcoder];
+        uint256 lastRewardRound = t.lastRewardRound;
 
-        EarningsPool.Data storage earningsPool = t.earningsPoolPerRound[_round];
+        EarningsPool.Data storage earningsPool = t.earningsPoolPerRound[currentRound];
+        EarningsPool.Data storage prevEarningsPool = t.earningsPoolPerRound[currentRound.sub(1)];
 
-        // if transcoder hasn't called 'reward()' for '_round' its 'transcoderFeeShare' and 'transcoderRewardCut'
-        // on the 'EarningsPool' for '_round' would not be initialized and the fee distribution wouldn't happen as expected
-        if (_round > t.lastRewardRound) {
+        // if transcoder hasn't called 'reward()' for '_round' its 'transcoderFeeShare', 'transcoderRewardCut' and 'totalStake'
+        // on the 'EarningsPoolV2' for '_round' would not be initialized and the fee distribution wouldn't happen as expected
+        // for cumulative fee calculation this would result in division by zero. 
+        if (currentRound > lastRewardRound) {
             earningsPool.setCommission(
                 t.rewardCut,
                 t.feeShare
             );
+            
+            earningsPool.setStake(t.earningsPoolPerRound[t.lastActiveStakeUpdateRound].totalStake);
+
+            t.activeCumulativeRewards = t.cumulativeRewards;
         }
 
+        // LIP-36: Add fees for the current round instead of '_round'
+        // https://github.com/livepeer/LIPs/issues/35#issuecomment-673659199
+        uint256 totalStake = earningsPool.totalStake;
+        if ( prevEarningsPool.cumulativeRewardFactor == 0 && lastRewardRound > roundsManager().LIPUpgradeRounds(36)) {
+            // if transcoder didn't call reward for 'currentRound - 1' nor 'currentRound', use the cumulativeRewardFactor for 'lastRewardRound'
+            if (lastRewardRound < currentRound) {
+                prevEarningsPool.cumulativeRewardFactor = t.earningsPoolPerRound[lastRewardRound].cumulativeRewardFactor;
+            } else {
+                // if transcoder called reward for 'currentRound' but not for 'currentRound - 1' (missed reward call)
+                // retroactively calculate what its cumulativeRewardFactor would have been for 'currentRound - 1' (cfr. previous lastRewardRound for transcoder)  
+                // based on rewards for currentRound
+                uint256 rewards = MathUtils.percOf(minter().currentMintableTokens(), totalStake, currentRoundTotalActiveStake);
+                uint256 transcoderCommissionRewards = MathUtils.percOf(rewards, earningsPool.transcoderRewardCut);
+                uint256 delegatorRewards = rewards.sub(transcoderCommissionRewards);
+                prevEarningsPool.cumulativeRewardFactor = MathUtils.percOf(earningsPool.cumulativeRewardFactor, MathUtils.percOf(delegatorRewards, totalStake)).add(MathUtils.percPoints(1,1));
```

Comment:
I tried this in a unit test, this doesn't seem quite correct either ?

Reply:
Not sure if this is the cause, but I also noticed this line:

```solidity
uint256 rewards = MathUtils.percOf(minter().currentMintableTokens(), totalStake, currentRoundTotalActiveStake);
```

`currentMintableTokens()` will decrease whenever reward is called (the rewards minted is subtracted). I think the value we want to use here is the total mintable tokens at the beginning of the round which can be expressed as:

```solidity
minter().currentMintableTokens().add(minter().currentMintedTokens())
```

which adds the remaining mintable tokens in the round with the tokens already minted in the round.

Reply:
We should add that definitely but it's unrelated to the `claimEarnings` tests in the BondingManager unit tests.

Reply:
The formula you portray seems correct but I don't think the solidity 'translation' is correct, could you confirm? 

It yields wildly inaccurate results for previously passing unit tests.

Reply:
I think we are also missing the `transcoderRewardStakeRewards` here to subtract from `delegatorsFees` , no ?

Reply:
> The formula you portray seems correct but I don't think the solidity 'translation' is correct, could you confirm?

Ah oof not sure what I was thinking there...

I think the following should be correct:

```solidity
MathUtils.percOf(
    earningsPool.cumulativeRewardFactor,
    totalStake,
    delegatorRewards.add(totalStake)
);
```

This should be equivalent to:

```
cumulativeRewardFactor * (1 / (1 + delegatorRewards / totalStake)) = ...
cumulativeRewardFactor * (1 / (delegatorRewards + totalStake / totalStake)) = ...
cumulativeRewardFactor * (totalStake / delegatorRewards + totalStake)
```

> I think we are also missing the transcoderRewardStakeRewards here to subtract from delegatorsFees , no ?

I don't think `transcoderRewardStakeRewards` is relevant here because those aren't subtracted from `delegatorRewards` in `updateTranscoderWithRewards()`. `transcoderRewardStakeRewards` represents the rewards earned by the cumulative rewards that the transcoder earned based on the portion of its self-delegated stake that comes from its `rewardCut`. The rest of the transcoder's self-delegated stake can earn a portion of `delegatorRewards`.

> but it's unrelated to the claimEarnings tests in the BondingManager unit tests.

The failing claimEarnings tests seem to be due to the actual values for calculated rewards and fees being slightly off from expected values for rewards and/or fees. For example, the acceptable delta between expected and actual rewards/fees in one of the tests is 2, but the actual delta is 4. I'm not sure if this is ok or not yet - will have to look into it more. It might be the case that the difference is the result of using a different method to calculate the expected result than the method used to calculate the actual result.

I also noticed that one of the pendingFees tests was failing [here](https://app.circleci.com/pipelines/github/livepeer/protocol/89/workflows/a1a9bbae-4ce8-41e0-9ce4-43f6cef4f2e2/jobs/1636). The failing test case:

```
5) Contract: BondingManager
       pendingFees
         should return pending fees when transcoder hasn't called reward for the previous round but has for the current round:

      AssertionError: expected '382' to equal '411'
      + expected - actual

      -382
      +411
      
      at _callee209$ (test/unit/BondingManager.js:7077:50)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:65:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:303:22)
      at Generator.prototype.(anonymous function) [as next] (node_modules/regenerator-runtime/runtime.js:117:21)
      at step (test/unit/BondingManager.js:25:191)
      at /home/circleci/protocol/test/unit/BondingManager.js:25:361
      at process._tickCallback (internal/process/next_tick.js:68:7)
```

I was able to fix that failing test case locally with the following changes:

```diff
diff --git a/contracts/bonding/BondingManager.sol b/contracts/bonding/BondingManager.sol
index 8273ba6..a11a3c7 100644
--- a/contracts/bonding/BondingManager.sol
+++ b/contracts/bonding/BondingManager.sol
@@ -349,7 +349,7 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
         // LIP-36: Add fees for the current round instead of '_round'
         // https://github.com/livepeer/LIPs/issues/35#issuecomment-673659199
         uint256 totalStake = earningsPool.totalStake;
-        if (prevEarningsPool.cumulativeRewardFactor == 0 && prevRewardRound == currentRound && prevRewardRound > roundsManager().LIPUpgradeRounds(36)) {
+        if (prevEarningsPool.cumulativeRewardFactor == 0 && lastRewardRound == currentRound) {
             // if transcoder called reward for 'currentRound' but not for 'currentRound - 1' (missed reward call)
             // retroactively calculate what its cumulativeRewardFactor would have been for 'currentRound - 1' (cfr. previous lastRewardRound for transcoder)
             // based on rewards for currentRound
@@ -358,7 +358,8 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
             uint256 delegatorRewards = rewards.sub(transcoderCommissionRewards);
             prevEarningsPool.cumulativeRewardFactor = MathUtils.percOf(
                 earningsPool.cumulativeRewardFactor,
-                MathUtils.percPoints(1, MathUtils.percPoints(1, 1).add(MathUtils.percPoints(delegatorRewards, totalStake)))
+                totalStake,
+                delegatorRewards.add(totalStake)
             );
         }

diff --git a/test/unit/BondingManager.js b/test/unit/BondingManager.js
index afb7483..8db67e7 100644
--- a/test/unit/BondingManager.js
+++ b/test/unit/BondingManager.js
@@ -2597,6 +2597,9 @@ contract("BondingManager", accounts => {
                 (pendingFees0 + pendingFees1 + pendingFees2).toString()
             )

+            await fixture.minter.setMockUint256(functionSig("currentMintableTokens()"), 0)
+            await fixture.minter.setMockUint256(functionSig("currentMintedTokens()"), 1000)
+            await fixture.roundsManager.execute(bondingManager.address, functionSig("setCurrentRoundTotalActiveStake()"))
             await fixture.roundsManager.setMockUint256(functionSig("currentRound()"), currentRound + 4)
             await bondingManager.reward({from: transcoder})
```

Notes on the above diff:

- The conditional for determining whether we need to retroactively calculate the cumulativeRewardFactor for the previous round is updated
    - `prevRewardRound == currentRound` -> `lastRewardRound == currentRound`
        - This is necessary because otherwise `prevRewardRound = currentRound - 1` so the previous check would always be false
    - The LIP upgrade round check is removed as well because I don't think it's necessary? If the previous round was not after the LIP upgrade round then I think the cumulativeRewardFactor for the previous round would just evaluate to `MathUtils(1, 1)` which should be fine...worth double checking this though
- The formula for retroactively calculating the cumulativeRewardFactor in the previous round is updated
- The test is updated with appropriate mock statements. Without those mock statements, incorrect values would be used for the retroactive calculation